### PR TITLE
Add getTransactionBodyContent

### PR DESCRIPTION
**Pull request by cblp**

### CODE REVIEW

File: cardano-api/src/Cardano/Api/TxBody.hs [hunk 1]
```diff
@@ -1336,41 +1376,309 @@ collectTxBodySimpleScripts TxBodyContent {
...
+
 toShelleySimpleScript :: SimpleScriptInEra era
                       -> Ledger.Script (ShelleyLedgerEra era)
 toShelleySimpleScript (SimpleScriptInEra langInEra version script) =
     toShelleyScript (ScriptInEra langInEra (SimpleScript version script))
 
-toShelleyWithdrawal :: [(StakeAddress, Lovelace, a)] -> Shelley.Wdrl StandardCrypto
+
+getShelleyTxBodyContent ::  Shelley.TxBody (ShelleyLedgerEra ShelleyEra)
+                        ->  Maybe
```
Comment (Jimbo4350) [inline cardano-api/src/Cardano/Api/TxBody.hs:1387]: Formatting. Also why do we need ` Maybe (Shelley.Metadata (ShelleyLedgerEra ShelleyEra))`? This function should also be parameterized by `ShelleyBasedEra era` so  that we don't need to have a separate one for each era.

File: cardano-api/src/Cardano/Api/TxBody.hs [hunk 2]
```diff
@@ -1336,41 +1376,309 @@ collectTxBodySimpleScripts TxBodyContent {
...
+
 -- | In the Allegra and Mary eras the auxiliary data consists of the tx metadata
 -- and the axiliary scripts.
 --
-toAllegraAuxiliaryData :: forall era ledgeera.
-                          ShelleyLedgerEra era ~ ledgeera
-                       => Ledger.AuxiliaryData ledgeera ~ Allegra.AuxiliaryData ledgeera
-                       => Ledger.AnnotatedData (Ledger.Script ledgeera)
-                       => Ord (Ledger.Script ledgeera)
+toAllegraAuxiliaryData :: forall era ledgerEra.
```
Comment (Jimbo4350) [inline cardano-api/src/Cardano/Api/TxBody.hs:1650]: Lets follow the pattern already established in cardano-api. 

`ledgerEra` should be `ledgerera`

File: cardano-api/src/Cardano/Api/TxBody.hs [hunk 3]
```diff
@@ -1336,41 +1376,309 @@ collectTxBodySimpleScripts TxBodyContent {
...
+
 -- | In the Shelley era the auxiliary data consists only of the tx metadata
 toShelleyAuxiliaryData :: Map Word64 TxMetadataValue
                        -> Ledger.AuxiliaryData StandardShelley
 toShelleyAuxiliaryData m =
     Shelley.Metadata
       (toShelleyMetadata m)
 
+
+fromShelleyAuxiliaryData  :: Ledger.AuxiliaryData StandardShelley
```
Comment (Jimbo4350) [inline cardano-api/src/Cardano/Api/TxBody.hs:1642]: This can be parameterized with `ShelleyBasedEra era` to avoid repetition

File: cardano-api/src/Cardano/Api/TxBody.hs [hunk 4]
```diff
@@ -1381,24 +1689,38 @@ makeByronTransaction :: [TxIn]
                      -> [TxOut ByronEra]
                      -> Either (TxBodyError ByronEra) (TxBody ByronEra)
 makeByronTransaction txIns txOuts =
```
Comment (Jimbo4350) [inline cardano-api/src/Cardano/Api/TxBody.hs:1691]: I wouldn't bother refactoring this as it's deprecated

**Review by Jimbo4350 — CHANGES_REQUESTED** [review]
Nice work, some comments below.

File: cardano-api/src/Cardano/Api/TxBody.hs [hunk 5]
```diff
@@ -1171,7 +1211,7 @@ makeShelleyTransactionBody era@ShelleyBasedEraAllegra
 
     return $
       ShelleyTxBody era
-        (Allegra.TxBody
+        (ShelleyMA.TxBody
```
[the remark this answers is not in the document]
Comment (cblp) [reply cardano-api/src/Cardano/Api/TxBody.hs:1214]: Do you offer to use the Allegra prefix for Mary functions or add an identical Mary prefix?

File: cardano-api/src/Cardano/Api/TxBody.hs [hunk 6]
```diff
@@ -1336,41 +1376,309 @@ collectTxBodySimpleScripts TxBodyContent {
...
+      case update of
+        SNothing -> TxUpdateProposalNone
+        SJust p ->
+          TxUpdateProposal UpdateProposalInMaryEra $ fromShelleyUpdate p
+    txMintValue
+      | isZero mint = TxMintNone
+      | otherwise = TxMintValue MultiAssetInMaryEra (fromMaryValue mint) ViewTx
+
+
+toShelleyWithdrawal :: [(StakeAddress, Lovelace, build)]
```
[the remark this answers is not in the document]
Comment (cblp) [reply cardano-api/src/Cardano/Api/TxBody.hs:1613]: It was a try to increase readability for myself, reverted.

File: cardano-api/src/Cardano/Api/TxBody.hs [hunk 7]
```diff
@@ -1336,41 +1376,309 @@ collectTxBodySimpleScripts TxBodyContent {
...
+    txUpdateProposal =
+      case Shelley._txUpdate body of
+        SNothing -> TxUpdateProposalNone
+        SJust p ->
+          TxUpdateProposal UpdateProposalInShelleyEra $ fromShelleyUpdate p
+
+
+getAllegraTxBodyContent ::  ShelleyMA.TxBody (ShelleyLedgerEra AllegraEra)
+                        ->  Maybe
+                              (ShelleyMA.AuxiliaryData
```
[the remark this answers is not in the document]
Comment (cblp) [reply cardano-api/src/Cardano/Api/TxBody.hs:1453]: This is much better! I didn't know this is allowed. I tried to follow the file's style.

File: cardano-api/src/Cardano/Api/TxBody.hs [hunk 8]
```diff
@@ -1054,6 +1072,21 @@ makeTransactionBody =
...
 
 
+getTransactionBodyContent ::  TxBody era
+                          ->  Either
+                                (TxBodyError era)
+                                (TxBodyContent ViewTx era)
+getTransactionBodyContent = \case
+  ByronTxBody body ->
+    Right $ getByronTxBodyContent body
+  ShelleyTxBody ShelleyBasedEraShelley body _scripts metadata ->
```
[the remark this answers is not in the document]
Comment (cblp) [reply cardano-api/src/Cardano/Api/TxBody.hs:1082]: Why? This specific aux is used for metadata only.

File: cardano-api/src/Cardano/Api/TxBody.hs (hunk 1 above)
Comment (cblp) [reply cardano-api/src/Cardano/Api/TxBody.hs:1387]: 1. This parameter is used to fill the `txMetadata` field.
2. These functions seem pretty different. Okay, I'll try to generalize.

File: cardano-api/src/Cardano/Api/TxBody.hs (hunk 2 above)
Comment (cblp) [reply cardano-api/src/Cardano/Api/TxBody.hs:1650]: done

File: cardano-api/src/Cardano/Api/TxBody.hs (hunk 3 above)
Comment (cblp) [reply cardano-api/src/Cardano/Api/TxBody.hs:1642]: Which repetition? Could you name specific functions?

File: cardano-api/src/Cardano/Api/TxBody.hs (hunk 4 above)
Comment (cblp) [reply cardano-api/src/Cardano/Api/TxBody.hs:1691]: I needed to extract `makeByronTransactionBodyContent`

File: cardano-api/src/Cardano/Api/TxBody.hs (hunk 3 above)
Comment (cblp) [reply cardano-api/src/Cardano/Api/TxBody.hs:1633]: Generalization of `fromAllegraAuxiliaryData` and `fromMaryAuxiliaryData` incurred such strange constraints:

```hs
fromAllegraBasedAuxiliaryData
  ::  forall era ledgerera
  .   ( ShelleyLedgerEra era ~ ledgerera
      , Ledger.AuxiliaryData ledgerera ~ ShelleyMA.AuxiliaryData ledgerera
      , FromCBOR (CBOR.Annotator (Ledger.Script ledgerera))
      , ToCBOR (Ledger.Script ledgerera)
      , Ord (Ledger.Script ledgerera)
      )
  =>  ShelleyBasedEra era
  ->  Ledger.AuxiliaryData (ShelleyLedgerEra era)
  ->  (Map Word64 TxMetadataValue, [ScriptInEra era])
fromAllegraBasedAuxiliaryData era (ShelleyMA.AuxiliaryData ms ss) =
  (fromShelleyMetadata ms, fromShelleyBasedScript era <$> toList ss)
```

File: cardano-api/src/Cardano/Api/TxBody.hs (hunk 5 above)
[the remark this answers is not in the document]
Comment (gandawa) [reply cardano-api/src/Cardano/Api/TxBody.hs]: I'm suggesting to leave it as it was. Let's have a discussion on slack about what the import naming should be.

[the remark this answers is not in the document]
Comment (Jimbo4350) [reply cardano-api/src/Cardano/Api/TxBody.hs]: I'm suggesting to leave it as it was. Let's have a discussion on slack about this.

File: cardano-api/src/Cardano/Api/TxBody.hs (hunk 1 above)
Comment (Jimbo4350) [reply cardano-api/src/Cardano/Api/TxBody.hs]: @cblp I think I understand why you are supplying the tx metadata in the function arguments. We shouldn't do this. The txbody stores the auxiliary data hash, we should display that instead.

File: cardano-api/src/Cardano/Api/TxBody.hs (hunk 3 above)
Comment (Jimbo4350) [reply cardano-api/src/Cardano/Api/TxBody.hs:1632]: This is fine for now.

File: cardano-api/src/Cardano/Api/TxBody.hs (hunk 1 above)
Comment (cblp) [reply cardano-api/src/Cardano/Api/TxBody.hs]: 1. Removed txMetadata logic. I don't understand why it is a problem.
2. `metadata` parameter is still used to check hash.
3. Should I add metadata hash display now?

File: cardano-api/src/Cardano/Api/TxBody.hs (hunk 5 above)
[the remark this answers is not in the document]
Comment (cblp) [reply cardano-api/src/Cardano/Api/TxBody.hs]: I reverted the old code. Is it good now?

File: cardano-api/src/Cardano/Api/TxBody.hs (hunk 3 above)
Comment (cblp) [reply cardano-api/src/Cardano/Api/TxBody.hs]: Applied.

**Comment by Jimbo4350** [comment]
@cblp So following the style of `cardano-api` this is what we are aiming for. You're heading in the right direction but we can further refactor to the following:

```
getTxBodyContent :: TxBody era -> TxBodyContent ViewTx era
getTxBodyContent (ByronTxBody annot) = getByronTxBodyContent annot
getTxBodyContent (ShelleyTxBody sbe txBody scripts mAuxData) =
  fromLedgerTxBody sbe txBody

fromLedgerTxBody
  :: ShelleyBasedEra era
  -> Ledger.TxBody (ShelleyLedgerEra era)
  -> TxBodyContent ViewTx era
fromLedgerTxBody sbe txBod =
  TxBodyContent
    { txIns = fromLedgerTxIns sbe txBod
    , txOuts = error ""
    , txFee = error ""
    , txValidityRange = error ""
    , txMetadata = error ""
    , txAuxScripts = error ""
    , txWithdrawals = error ""
    , txCertificates = error ""
    , txUpdateProposal = error ""
    , txMintValue = error ""
    }

fromLedgerTxIns
  :: ShelleyBasedEra era
  -> Ledger.TxBody (ShelleyLedgerEra era)
  -> [(TxIn,BuildTxWith ViewTx (Witness WitCtxTxIn era))]
fromLedgerTxIns sbe ledTxBody =
  case sbe of
    ShelleyBasedEraShelley ->
      [ (fromShelleyTxIn input, ViewTx)
      | input <- toList $ Shelley._inputs ledTxBody
      ]
    ShelleyBasedEraAllegra ->
      [ (fromShelleyTxIn input, ViewTx)
      | input <- toList $ Allegra.inputs' ledTxBody
      ]
    ShelleyBasedEraMary ->
      [ (fromShelleyTxIn input, ViewTx)
      | input <- toList $ Mary.inputs' ledTxBody
      ]
```
We can also remove the error checking as we already do that when constructing a `TxBody era` i.e :
```
makeShelleyTransactionBody :: ShelleyBasedEra era
                           -> TxBodyContent BuildTx era
                           -> Either (TxBodyError era) (TxBody era)                       
```

**Comment by cblp** [comment]
restored txMetadata code

**Comment by cblp** [comment]
@Jimbo4350 transposed code by field-then-era

File: cardano-api/src/Cardano/Api/TxBody.hs [hunk 9]
```diff
@@ -1184,10 +1478,10 @@ makeShelleyTransactionBody era@ShelleyBasedEraAllegra
              TxFeeImplicit era'  -> case era' of {}
              TxFeeExplicit _ fee -> toShelleyLovelace fee)
           (Allegra.ValidityInterval {
-             Allegra.invalidBefore    = case lowerBound of
```
Comment (Jimbo4350) [inline cardano-api/src/Cardano/Api/TxBody.hs:1187]: This is the Allegra era, so we should leave `invalidBefore` and `invalidHereafter` qualified

File: cardano-api/src/Cardano/Api/TxBody.hs [hunk 10]
```diff
@@ -1054,6 +1073,274 @@ makeTransactionBody =
...
+      )
+      where
+        Mary.ValidityInterval{invalidBefore, invalidHereafter} = Mary.vldt' body
+
+
+fromLedgerTxAuxiliaryData
+  :: ShelleyBasedEra era
+  -> Maybe (Ledger.AuxiliaryData (ShelleyLedgerEra era))
+  -> (TxMetadataInEra era, TxAuxScripts era)
+fromLedgerTxAuxiliaryData _ Nothing = (TxMetadataNone, TxAuxScriptsNone)
```
Comment (Jimbo4350) [inline cardano-api/src/Cardano/Api/TxBody.hs:1208]: This can be further simplified with:
```

fromLedgerAuxiliaryData
  :: ShelleyBasedEra era
  -> Ledger.AuxiliaryData (ShelleyLedgerEra era)
  -> (Map Word64 TxMetadataValue, [ScriptInEra era])
fromLedgerAuxiliaryData sbe aux =
  case sbe of
    ShelleyBasedEraShelley ->
      let Shelley.Metadata md = aux
      in (fromShelleyMetadata md, [])
    ShelleyBasedEraAllegra ->
      let Allegra.AuxiliaryData md sc = aux
      in (fromShelleyMetadata md, fromShelleyBasedScript sbe <$> toList sc)
    ShelleyBasedEraMary ->
      let Mary.AuxiliaryData md sc = aux
      in (fromShelleyMetadata md, fromShelleyBasedScript sbe <$> toList sc)
  ```

Comment (Jimbo4350) [reply cardano-api/src/Cardano/Api/TxBody.hs:1208]: We will be able to remove: `fromShelleyAuxiliaryData` and `fromAllegraBasedAuxiliaryData`

File: cardano-api/src/Cardano/Api/TxBody.hs [hunk 11]
```diff
@@ -1336,41 +1651,56 @@ collectTxBodySimpleScripts TxBodyContent {
...
 
+
 toShelleySimpleScript :: SimpleScriptInEra era
                       -> Ledger.Script (ShelleyLedgerEra era)
 toShelleySimpleScript (SimpleScriptInEra langInEra version script) =
     toShelleyScript (ScriptInEra langInEra (SimpleScript version script))
 
-toShelleyWithdrawal :: [(StakeAddress, Lovelace, a)] -> Shelley.Wdrl StandardCrypto
+
+toShelleyWithdrawal :: [(StakeAddress, Lovelace, a)]
```
Comment (Jimbo4350) [inline cardano-api/src/Cardano/Api/TxBody.hs:1661]: Unnecessary formatting change here

**Review by Jimbo4350 — APPROVED** [review]
Nice work! LGTM

**Comment by cblp** [comment]
bors merge
